core/src/main/java/com/linecorp/armeria/server/RouteCache.java

```diff
@@ -108,14 +145,34 @@ private RouteCache() {}
 
             final Routed<V> result = delegate.find(routingCtx);
             if (result.isPresent()) {
-                cache.put(routingCtx, result.value());
+                findCache.put(routingCtx, result.value());
             }
             return result;
         }
 
+        @Override
+        public Stream<Routed<V>> findAll(RoutingContext routingCtx) {
+            final List<V> cachedList = findAllCache.getIfPresent(routingCtx);
+            if (cachedList != null) {
+                return cachedList.stream().map(cached -> {
+                    final Route route = routeResolver.apply(cached);
+                    final RoutingResult routingResult = route.apply(routingCtx);
+                    return Routed.of(route, routingResult, cached);
+                });
+            }
+
+            final List<Routed<V>> result = delegate.findAll(routingCtx).collect(toImmutableList());
+            final List<V> valid = result.stream()
+                                        .filter(Routed::isPresent)
+                                        .map(Routed::value)
+                                        .collect(toImmutableList());
+            findAllCache.put(routingCtx, valid);
+            return result.stream();
```

Comment:
Maybe worth a comment, I'm a bit confused why we return the "cached" value (the one we just put into the cache) as is here, but have to manipulate it when reading from the cache, I don't think this is a normal pattern for a cache.

Reply:
> I'm a bit confused why we return the "cached" value (the one we just put into the cache) as is here

IIUC, `result` is not the cached value, the `valid`(manipulated value) is put into the cache.

> but have to manipulate it when reading from the cache, I don't think this is a normal pattern for a cache.

I think the original developer wants to share the cache instance so that I have to manipulate cached value. I also agree this is not a normal pattern for a cache.

If we need to change the cache strategy here, I think we need a sperate issue. WDYT?

Reply:
Leaving it is fine with me